Tutorials/Creating-an-XML-Site-Map/index.md

```diff
@@ -0,0 +1,322 @@
+
+### Rendering a site map entry
+
+We will retrieve each page in the site as **IPublishedContent**, we're not really concerned with any of the different types of properties a page may have, for the site map, we're only interested in the Url, when it was last modified etc, whether it was hidden, but you could write this with Modelsbuilder too (should this example use modelsbuilder? hmmm) - as we'll be writing out this url markup for an IPublishedContent item in different places as we loop to build the sitemap, let's create a Razor Helper that we can pass the IPublishedContent object to, to be responsible for rendering the markup.
+
+```chsarp
+@helper RenderSiteMapUrlEntry(IPublishedContent node)
+{
+    //we're passing 'true' as an additional parameter to GetPropertyValue to read the value from the parent nodes, so this setting could be set 'per section
+    var changeFreq = node.GetPropertyValue<string>("searchEngineChangeFrequency", true);
+    if (String.IsNullOrEmpty(changeFreq))
+    {
+        changeFreq = "monthly";
+    }
+    //with the relative priority, this is a per page setting, so we're not passing true
+    var priority = node.GetPropertyValue<string>("searchEngineRelativePriority");
+    if (String.IsNullOrEmpty(priority))
+    {
+        priority = "0.5";
+    }
+    <url>
+        <loc>@EnsureUrlStartsWithDomain(node.UrlWithDomain())</loc>
+        <lastmod>@(string.Format("{0:s}+00:00", node.UpdateDate))</lastmod>
+        <changefreq>@changeFreq</changefreq>
+        <priority>@priority</priority>
+    </url>
+}
+```
+
+#### EnsureUrlStartsWithDomain - Razor Function
+
+We'll create a Razor Function to handle ensuring the urls on our sitemap have the correct domain (you aren't meant to have relative urls on a sitemap - citation needed), 
+
+Razor functions exist inside a single @functions block inside your template
+
+```chsarp
+@functions {
+    private static string EnsureUrlStartsWithDomain(string url)
+    {
+        if (url.StartsWith("http")){
+            return url;
+        }
+    else {
+        //whatever makes sense for your site here...
+        var domainPrefix = string.Format("https://{0}/", HttpContext.Current.Request.ServerVariables["HTTP_HOST"]);
+        return domainPrefix + url;
+    }}
+}
+```
+
+#### Xml Sitemap for the homepage
+
+So for the homepage we'll now have:
+
+    @inherits Umbraco.Web.Mvc.UmbracoTemplatePage
+    @{Layout = null; Response.ContentType = "text/xml";IPublishedContent siteHomePage = Model.Content.Site(); }<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemalocation="http://www.google.com/schemas/sitemap/0.9 http://www.sitemaps.org/schemas/sitemap/0.9/sitemap.xsd" xmlns:image="http://www.google.com/schemas/sitemap-image/1.1">@RenderSiteMapUrlEntry(siteHomePage)</urlset>
+
+and this will render a sitemap entry for the homepage, which isn't very comprehensive!
+
```

Comment:
Once I got to this point, for some reason the Model.Content.Site() would return the XmlSiteMap node I had created in root, even though I had a home node there already. Not too sure why, do I need to do anything extra? 🙂

My setup was like this:
![udklip](https://user-images.githubusercontent.com/36075913/48624764-8c76a580-e9ad-11e8-97c3-919ae3ee62fd.PNG)